aggregator/handler/sharded.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package handler
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/m3db/m3aggregator/aggregator"
+	"github.com/m3db/m3aggregator/sharding"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errEmptyBackendServerShards = errors.New("empty backend server shards")
+)
+
+// BackendServerShard contains a backend server shard.
+type BackendServerShard struct {
+	// Name of this shard for logging and metrics.
+	Name string `yaml:"name"`
+
+	// Shard range.
+	Range sharding.ShardSet `yaml:"range" validate:"nonzero"`
+
+	// Servers owning the shard.
+	Servers []string `yaml:"servers" validate:"nonzero"`
+}
+
+type shardedHandlerMetrics struct {
+	shardNotAssigned tally.Counter
+}
+
+func newShardedHandlerMetrics(scope tally.Scope) shardedHandlerMetrics {
+	return shardedHandlerMetrics{
+		shardNotAssigned: scope.Counter("shard-not-assigned"),
+	}
+}
+
+type shardedHandler struct {
+	handlers []aggregator.Handler
+	metrics  shardedHandlerMetrics
+}
+
+// NewShardedHandler creates a new sharded handler.
```

Comment:
Just for my own understanding, how do we handle updates to the shard topology, i.e. backends shards being added or removed? Seems it's just static and defined in configuration now, so I wanted to confirm.

Reply:
That's not supported at the moment, the backend topology is statically configured. That said at some point in the future we'll change the backend to read the topology from KV and route things dynamically based on KV updates to avoid deploying the agg tier when a backend server goes down.